amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/HierarchicalShardSyncer.java

```diff
@@ -918,6 +936,21 @@ public synchronized List<Lease> determineNewLeasesToCreate(final List<Shard> sha
                     .map(streamId -> streamId.serialize()).orElse("");
             final Set<String> shardIdsOfCurrentLeases = currentLeases.stream()
                     .peek(lease -> log.debug("{} : Existing lease: {}", streamIdentifier, lease))
+                    .filter(lease -> {
+                        if (lease instanceof MultiStreamLease) {
+                            if (StreamProcessingMode.SINGLE_STREAM_COMPATIBLE_MODE == multiStreamArgs.streamProcessingMode()
+                                || StreamProcessingMode.SINGLE_STREAM_UPGRADE_MODE == multiStreamArgs.streamProcessingMode()) {
+                                MultiStreamLease multiStreamLease = (MultiStreamLease) lease;
+                                // In compatibility and upgrade mode we make sure that the stream identifier in configuration
+                                // matches the lease stream identifier.
+                                return multiStreamArgs.streamIdentifier.serialize().equals(multiStreamLease.streamIdentifier());
```

Comment:
i want to avoid these conditional checks. Can we just filter by given streamId, if the filtered out lease is not empty then we log an error and just ignore those leases. or it maybe better to have a new method to scan and filter for an absence of the attribute or presence with a given value. so everything else if present will be ignored.

Reply:
We cannot ignore the single stream leases here which have no streamIdentifier.